### PR TITLE
enhance imageRegistry API to support multi-registries

### PR DESCRIPTION
**Pull request by zhiweiyin318**
Signed-off-by: Zhiwei Yin <zyin@redhat.com>

issue: https://github.com/stolostron/backlog/issues/21706

1. add registries filed to spec:
```yaml 
apiVersion: imageregistry.open-cluster-management.io/v1alpha1
kind: ManagedClusterImageRegistry
metadata:
  name: custom-image-registry
  namespace: default
spec:
  placementRef:
    group: cluster.open-cluster-management.io
    name: custom-placement
    resource: placements
  pullSecret:
    name: custom-pull-secret
  registry: quay.io/rhacm2
  registries:
    - mirror: quay.io/rhacm2/
      source: registry.redhat.io/rhacm2
    - mirror: quay.io/multicluster-engine
      source: registry.redhat.io/multicluster-engine
```
to compatible with 2.4,  registry filed will be used when registries is empty, which means the mirror is the registry but the source is all .

2. add annotations to cluster
```
open-cluster-management.io/image-registries: '{"pullSecret":"namespace.pullSecret", "registries":[{"mirror":"quay.io/rhacm2", "source":"registry.redhat.io/rhacm2"},{"mirror":"quay.io/multicluster-engine", "source":"registry.redhat.io/multicluster-engine"}]}'
```
3. keep the `imageRegistryLabel` of cluster to handle the case the cluster is removed from a placementDecision.
need to remove the label and annotation when the cluster is removed from a placementDecision but has `imageRegistryLabel`.

5. update the helpers to override images registry from the annotation.

### CODE REVIEW

File: pkg/controllers/imageregistry/imageregistry_controller.go [hunk 1]
```diff
@@ -218,37 +219,35 @@ func (r *Reconciler) cleanImageRegistryLabel(ctx context.Context, namespace, ima
 	return nil
 }
 
-// updateImageRegistryLabel will update the label 'open-cluster-management.io/image-registry: namespace.imageRegistry' of clusters.
-// 1. remove the label if the labeled cluster is not selected.
-// 2. update the label of selected clusters.
-// 2.1 add the label if the selected cluster has not the ClusterImageRegistryLabel.
```
Comment (qiujian16) [inline pkg/controllers/imageregistry/imageregistry_controller.go:224]: has not --> does not have

File: pkg/controllers/imageregistry/imageregistry_controller.go [hunk 2]
```diff
@@ -218,37 +219,35 @@ func (r *Reconciler) cleanImageRegistryLabel(ctx context.Context, namespace, ima
...
+// 2.1 add the label and annotations if the selected cluster has not the ClusterImageRegistryLabel.
 // 2.2 send warning event if the selected cluster has another ClusterImageRegistryLabel.
-func (r *Reconciler) updateImageRegistryLabel(ctx context.Context, selectedClusters []string, namespace, imageRegistry string) error {
+func (r *Reconciler) updateAllClustersImageRegistry(ctx context.Context,
+	selectedClusters []string, imageRegistry *v1alpha1.ManagedClusterImageRegistry) error {
 	var errs []error
 	needUpdateClustersMap := sets.NewString(selectedClusters...)
 
-	imageRegistryLabelValue := generateImageRegistryLabelValue(namespace, imageRegistry)
+	imageRegistryLabelValue := generateImageRegistryLabelValue(imageRegistry.Namespace, imageRegistry.Name)
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/controllers/imageregistry/imageregistry_controller.go:232]: keep the imageRegistryLabel of cluster to handle the case the cluster is removed from a placementDecision.
need to remove the label and annotation when the cluster is removed from a placementDecision but has imageRegistryLabel.

File: pkg/controllers/imageregistry/imageregistry_controller.go [hunk 3]
```diff
@@ -384,3 +391,31 @@ func (r *Reconciler) updateImageRegistryCondition(ctx context.Context, namespace
...
+
+	if len(imageRegistry.Spec.Registries) != 0 {
+		annotationRegistries, _ := json.Marshal(imageRegistry.Spec.Registries)
+		return string(annotationRegistries)
+	}
+
+	registries := []v1alpha1.Registries{
+		{
+			Mirror: imageRegistry.Spec.Registry,
+			Source: "",
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/controllers/imageregistry/imageregistry_controller.go:416]: the source is from the image name in manifest configmap created by mch operator. for example registry is `quay.io:443/acm-d` in e2e test env.

File: pkg/controllers/imageregistry/imageregistry_controller.go [hunk 4]
```diff
@@ -311,7 +312,13 @@ func (r *Reconciler) updateClusterImageRegistryLabel(ctx context.Context, cluste
 		}
 
 		modified := false
-		utils.MergeMap(&modified, &cluster.Labels, map[string]string{v1alpha1.ClusterImageRegistryLabel: imageRegistryLabelValue})
+		utils.MergeMap(&modified, &cluster.Annotations,
+			map[string]string{
+				v1alpha1.AnnotationRegistries: getAnnotationRegistries(imageRegistry),
+				v1alpha1.AnnotationPullSecret: getAnnotationPullSecret(imageRegistry)})
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/controllers/imageregistry/imageregistry_controller.go:318]: how about 
```yaml
open-cluster-management.io/registries: '{"pullSecret":"namespace.pullSecret", "registries":[{"mirror":"quay.io/rhacm2", "source":"registry.redhat.io/rhacm2"},{"mirror":"quay.io/multicluster-engine", "source":"registry.redhat.io/multicluster-engine"}]}'
```

[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/controllers/imageregistry/imageregistry_controller.go:318]: it is ok to me

File: pkg/controllers/imageregistry/imageregistry_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/controllers/imageregistry/imageregistry_controller.go:416]: hrm, so there is an implication when source is empty. we should doc it.

File: pkg/apis/imageregistry/v1alpha1/imageregistry_types.go [hunk 5]
```diff
@@ -23,6 +27,13 @@ type ImageRegistrySpec struct {
 	PlacementRef PlacementRef `json:"placementRef"`
 }
 
+type Registries struct {
+	// Mirror is the mirrored registry of the source
+	Mirror string `json:"mirror"`
```
Comment (qiujian16) [inline pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:32]: this is required?

File: pkg/apis/imageregistry/v1alpha1/imageregistry_types.go [hunk 6]
```diff
@@ -23,6 +27,13 @@ type ImageRegistrySpec struct {
...
+	// Source is the source registry
+	Source string `json:"source"`
```
Comment (qiujian16) [inline pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:34]: this is optional?

File: pkg/apis/imageregistry/v1alpha1/imageregistry_types.go (hunk 5 above)
Comment (zhiweiyin318) [reply pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:35]: yes. add doc for this filed.

File: pkg/apis/imageregistry/v1alpha1/imageregistry_types.go (hunk 6 above)
Comment (zhiweiyin318) [reply pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:39]: yes. add doc for this filed.

Comment (qiujian16) [reply pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:39]: what will be the source when this value is empty?

Comment (zhiweiyin318) [reply pkg/apis/imageregistry/v1alpha1/imageregistry_types.go:39]: as the same as using `Registry` only. the registry address path of all images will be replaced by the Mirror.

**Pull request merged**